src/snapred/backend/dao/ingredients/GroceryListItem.py

```diff
@@ -54,4 +58,20 @@ def validate_ingredients_for_groceries(cls, v):
                     raise ValueError("a grouping workspace requires an instrument source")
                 if v.get("instrumentSource") is None:
                     raise ValueError("a grouping workspace requries an instrument source")
+        # output (i.e. special-order) workspaces
+        if v["workspaceType"] == "diffcal_output":
+            if v.get("runNumber") is None:
+                raise ValueError("diffraction-calibration data output requires a run number")
+            if v.get("isOutput") == False:
+                raise ValueError("diffraction-calibration data output is special-order only")
```

Comment:
could these be moved to a helper method? Looks like they happen 3 time

Reply:
I think this is a situation where inheritance would help.  We really have a bunch if different kinds of grocery list items, one kind of item for neutron data, one for diffcal tables, one for grouping workspaces, etc.

Reply:
These are good comments.  I'd recommend that all of this wait until the next time this class gets refactored.  **My own "two cents" here is that it seems overly difficult to add new workspace types (unless someone is _very_ familiar with this class).  I'd prefer to see all of this automatically implemented from a YML file somewhere.**